node/lib/object_pool.js

```diff
@@ -0,0 +1,165 @@
+        assert(self.size === 0);
+    }
+    return self._previous === self && self._next === self;
+};
+
+DoublyLinkedList.prototype.insert = function insert(node) {
+    var self = this;
+    node._next = self;
+    node._previous = self._previous;
+    self._previous._next = node;
+    self._previous = node;
+    self.size++;
+};
+
+DoublyLinkedList.prototype.find = function find(node) {
+    var self = this;
+    var cur = self._next;
+    while (cur !== self) {
+        if (cur === node) return true;
+        cur = cur._next;
+    }
+
+    return false;
+};
+
+DoublyLinkedList.prototype.count = function count() {
+    var self = this;
+    var c = 0;
+    var cur = self._next;
+    while (cur !== self) {
+        c++;
+        cur = cur._next;
+    }
+
+    assert(c === self.size);
+    return c;
+};
+
+DoublyLinkedList.prototype.remove = function remove(node) {
+    var self = this;
+    node._previous._next = node._next;
+    node._next._previous = node._previous;
+    self.size--;
+};
+
+DoublyLinkedList.prototype.pop = function pop() {
+    var self = this;
+    if (self.isEmpty()) {
+        return null;
+    }
+
+    var node = self._next;
+    self.remove(node);
+    return node;
+};
+
+function ObjectPool(options) {
+    if (!(this instanceof ObjectPool)) {
+        return new ObjectPool(options);
+    }
+
+    var self = this;
+    self.name = options.name;
+    self.create = options.create;
+    self.staticPoolSize = options.staticPoolSize;
+    self.maxPoolSize = options.maxPoolSize;
+    if (self.maxPoolSize === undefined) {
+        self.maxPoolSize = 0xFFFFFFFF;
+    }
+
+    self.freeList = new DoublyLinkedList();
+    self.usedList = new DoublyLinkedList();
+
+    if (self.staticPoolSize) {
+        for (var i = 0; i < self.staticPoolSize; i++) {
+            var obj = self.create();
+            self.freeList.insert(obj);
+        }
+    }
+}
+
+ObjectPool.prototype.allocate = function allocate() {
+    var self = this;
+    var obj;
+
+    if (self.usedList.size >= self.maxPoolSize) {
+        obj = self.create.apply(null, arguments);
+        return obj;
+    }
+
+    // console.log('++++++', self.name, self.freeList.size, self.usedList.size);
+
+    if (self.freeList.isEmpty()) {
+        obj = self.create.apply(null, arguments);
+    } else {
+        obj = self.freeList.pop();
+        obj.setup.apply(obj, arguments);
+    }
+
+    self.usedList.insert(obj);
+
+    // console.log('------', self.name, self.freeList.size, self.usedList.size);
+
+    obj._inUse = true;
+    return obj;
+};
+
+ObjectPool.prototype.release = function release(obj) {
+    var self = this;
+
+    if (!obj._inUse) {
+        return;
+    }
+
+    // console.log('++++++', self.name, self.freeList.size, self.usedList.size);
+
+    assert(self.usedList.size > 0);
+    self.usedList.remove(obj);
+    self.freeList.insert(obj);
+
+    // console.log('------', self.name, self.freeList.size, self.usedList.size);
+
+    obj.clear();
```

Comment:
Actually we do not need to clear but we do need to `deallocate()`

The `deallocate()` method needs to only clear the most expensive fields to avoid memory leaks. Specifically removing references to buffers related to frames and requests.